source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -131,20 +131,39 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
                                     const Http::ResponseHeaderMap& response_headers,
                                     const ResponseMetadata& metadata) {
   const auto& simple_lookup_context = static_cast<const SimpleLookupContext&>(lookup_context);
-  const Key& key = simple_lookup_context.request().key();
+  const LookupRequest& request = simple_lookup_context.request();
+  const Key& key = request.key();
   absl::WriterMutexLock lock(&mutex_);
 
   auto iter = map_.find(key);
   if (iter == map_.end() || !iter->second.response_headers_) {
     return;
   }
-  auto& entry = iter->second;
+  Entry& firstEntry = iter->second;
 
-  // TODO(tangsaidi) handle Vary header updates properly
-  if (VaryHeaderUtils::hasVary(*(entry.response_headers_))) {
-    return;
+  if (VaryHeaderUtils::hasVary(*(firstEntry.response_headers_))) {
+    // sanity check whether the vary field is the same
+    ASSERT(VaryHeaderUtils::hasEqualVaryValues(*(firstEntry.response_headers_), response_headers));
+
+    // create vary identifier and varied_request_key
+    absl::btree_set<absl::string_view> vary_header_values =
+        VaryHeaderUtils::getVaryValues(response_headers);
+    const absl::optional<std::string> vary_identifier = VaryHeaderUtils::createVaryIdentifier(
+        request.varyAllowList(), vary_header_values, request.requestHeaders());
+
+    Key varied_request_key = request.key();
+    varied_request_key.add_custom_fields(vary_identifier.value());
+
+    iter = map_.find(varied_request_key);
```

Comment:
perf nit: this find() does not appear to provide any value when compiled for release. It's not obvious to me whether the compiler would optimize it away. Maybe use `#ifndef NDEBUG`, or bury this assert-testing logic, including the find() call, in a helper function which you could call in an ASSERT?

Reply:
The `iter` variable was declared in the function scope. It would overwrite the previously stored value of the iterator on line 138. I probably should add a comment about this to make it more obvious to readers. This is sort of necessary because the value we want to extract `Entry& entry` is a reference, which cannot be assigned.

The only alternative I can think of is to add a helper function that returns an `Entry&`. But since the logic is tightly coupled with this function, we would need to pass a lot of parameters to the helper function.

Reply:
oh sorry about that. I mis-read the snippet in github reviews and thought line 162 was the end of the function.